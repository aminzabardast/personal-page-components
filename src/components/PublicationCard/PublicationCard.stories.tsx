import React from 'react'
import { Meta, StoryObj } from '@storybook/react'
import PublicationCard from './PublicationCard'

const meta: Meta<typeof PublicationCard> = {
    title: 'Publications/Card',
    component: PublicationCard,
}

export default meta

type Story = StoryObj<typeof meta>

const bibtex: string =
    '@article{zabardast2023automated,\n' +
    '    title={An automated framework for evaluation of deep learning models for splice site predictions},\n' +
    '    author={Zabardast, Amin and Tamer, Elif G{\\"u}ney and Son, Ye{\\c{s}}im Ayd{\\i}n and Y{\\i}lmaz, Arif},\n' +
    '    journal={Scientific Reports},\n' +
    '    volume={13},\n' +
    '    number={1},\n' +
    '    pages={10221},\n' +
    '    year={2023},\n' +
    '    publisher={Nature Publishing Group UK London}\n' +
    '}'

const apa: string =
    'Zabardast, A., Tamer, E. G., Son, Y. A., & Yılmaz, A. (2023). An automated framework for evaluation of deep learning models for splice site predictions. Scientific Reports, 13(1), 10221.'

export const Example1: Story = {
    name: 'Citation',
    render: () => {
        return (
            <PublicationCard>
                <header>
                    <strong>Zabardast, A.</strong>, Tamer, E.G., Son, Y.A. and
                    Yılmaz, A., 2023.{' '}
                    <strong>
                        An automated framework for evaluation of deep learning
                        models for splice site predictions
                    </strong>
                    . Scientific Reports, 13(1), p.10221.
                </header>
                <header title="Bibtex">{bibtex}</header>
                <header title="APA">{apa}</header>
            </PublicationCard>
        )
    },
}
