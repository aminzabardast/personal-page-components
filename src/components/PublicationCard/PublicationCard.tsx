import React, { ReactNode, useState } from 'react'
import {
    Card,
    CardContent,
    Typography,
    Accordion,
    AccordionSummary,
    AccordionDetails,
    Stack,
} from '@mui/material'
import { get, isArray, filter, map } from 'lodash'

type ChildrenProp = ReactNode[]
type ChildProp = ReactNode

interface PublicationCardProp {
    children: ChildrenProp | ChildProp
}

const textToHtml = (value: string) => {
    // TODO: make this a different component with copy button and better scroll.
    // Check if build is minimal
    return (
        <pre
            dangerouslySetInnerHTML={{ __html: value }}
            style={{ overflowX: 'scroll' }}
        ></pre>
    )
}

const PublicationCard = ({ children }: PublicationCardProp) => {
    const [expanded, setExpanded] = useState(0)
    const handleChange = (panel: number) => {
        return (event: React.SyntheticEvent, isExpanded: boolean) => {
            setExpanded(isExpanded ? panel : 0)
        }
    }

    const citationBody: ChildrenProp | ChildProp = isArray(children)
        ? get(children, '0')
        : children
    const citationReferences: ChildrenProp = isArray(children)
        ? filter(children, (_, index) => index > 0)
        : []
    return (
        <Card elevation={0} variant="outlined">
            <CardContent>
                <Stack spacing={2} direction={'column'}>
                    <div data-testid="az-citation-body">{citationBody}</div>
                    <div data-testid="az-citation-references">
                        {map(
                            citationReferences,
                            (item: ChildProp, idx: number) => {
                                idx += 1
                                return (
                                    <Accordion
                                        key={idx}
                                        expanded={expanded === idx}
                                        onChange={handleChange(idx)}
                                    >
                                        <AccordionSummary>
                                            <Typography variant="body1">
                                                {get(item, 'props.title')}
                                            </Typography>
                                        </AccordionSummary>
                                        <AccordionDetails>
                                            {textToHtml(
                                                get(
                                                    item,
                                                    'props.children',
                                                    ''
                                                )
                                            )}
                                        </AccordionDetails>
                                    </Accordion>
                                )
                            }
                        )}
                    </div>
                </Stack>
            </CardContent>
        </Card>
    )
}

export default PublicationCard
