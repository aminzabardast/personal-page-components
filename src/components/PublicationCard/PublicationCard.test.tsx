import React from 'react'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import PublicationCard from './PublicationCard'
import renderer from 'react-test-renderer'

describe('<PublicationCard></PublicationCard>', () => {
    it('should render', () => {
        render(
            <PublicationCard>
                <div>A Citation.</div>
            </PublicationCard>
        )
    })
    it('should put the 1st child as the body of the citation', async () => {
        const citation: React.JSX.Element = (
            <header>
                Amin Zabardast, <strong>A Citation</strong>, <i>Year 2023</i>.
            </header>
        )
        render(<PublicationCard>{citation}</PublicationCard>)
        // Check if the element is inserted
        expect(
            screen.getByTestId('az-citation-body').querySelector('header')
        ).toBeInTheDocument()
        expect(screen.getByTestId('az-citation-body')).not.toBeEmptyDOMElement()
    })
    it('should put anything after the 1st child into an Accordion', async () => {
        const citation: React.JSX.Element = <header>Citation.</header>
        const bibtex: React.JSX.Element = (
            <header title="Bibtex">
                @article zabardast2023automated
                This is interesting
            </header>
        )
        const apa: React.JSX.Element = (
            <header title="APA">
                @article zabardast2023automated
                This is interesting
            </header>
        )
        render(
            <PublicationCard>
                {citation}
                {bibtex}
                {apa}
            </PublicationCard>
        )
        // Check if it is an Accordion
        expect(
            screen
                .getByTestId('az-citation-references')
                .querySelector('.MuiAccordion-root')
        ).toBeInTheDocument()
        // Check if the number of inserted elements are correct
        expect(
            screen.getByTestId('az-citation-references').querySelectorAll('pre')
                .length
        ).toEqual(2)
        // Check if titles are parsed and inserted correctly.
        expect(screen.getByText('Bibtex')).toBeInTheDocument()
        expect(screen.getByText('APA')).toBeInTheDocument()
    })
})
describe('<PublicationCard> SnapShot Test', () => {
    it('should keep producing consistent result.', () => {
        const tree = renderer
            .create(
                <PublicationCard>
                    <header>Citation.</header>
                    <header title="Bibtex">
                        @article zabardast2023automated
                        This is interesting
                    </header>
                </PublicationCard>
            )
            .toJSON()
        expect(tree).toMatchSnapshot()
    })
})
