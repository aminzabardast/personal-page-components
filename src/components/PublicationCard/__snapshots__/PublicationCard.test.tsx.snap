// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PublicationCard> SnapShot Test should keep producing consistent result. 1`] = `
<div
  className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <div
      className="MuiStack-root css-1sazv7p-MuiStack-root"
    >
      <div
        data-testid="az-citation-body"
      >
        <header>
          Citation.
        </header>
      </div>
      <div
        data-testid="az-citation-references"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1yzeyp9-MuiPaper-root-MuiAccordion-root"
        >
          <div
            aria-expanded={false}
            className="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-4pup92-MuiButtonBase-root-MuiAccordionSummary-root"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            <div
              className="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-18x7tpe-MuiAccordionSummary-content"
            >
              <p
                className="MuiTypography-root MuiTypography-body1 css-45zogx-MuiTypography-root"
              >
                Bibtex
              </p>
            </div>
          </div>
          <div
            className="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
            style={
              {
                "minHeight": "0px",
              }
            }
          >
            <div
              className="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
            >
              <div
                className="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
              >
                <div
                  className="MuiAccordion-region"
                  role="region"
                >
                  <div
                    className="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
                  >
                    <div
                      data-testid="citation-box"
                    >
                      <button
                        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                        disabled={false}
                        onBlur={[Function]}
                        onClick={[Function]}
                        onContextMenu={[Function]}
                        onDragLeave={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={0}
                        type="button"
                      >
                        <svg
                          aria-hidden={true}
                          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"
                          />
                        </svg>
                      </button>
                      <pre
                        dangerouslySetInnerHTML={
                          {
                            "__html": "@article zabardast2023automated This is interesting",
                          }
                        }
                        style={
                          {
                            "overflowX": "auto",
                          }
                        }
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
